overlord/snapstate/autorefresh_gating_test.go

```diff
@@ -211,6 +223,392 @@ slots:
+	mockInstalledSnap(c, st, snapByaml, false)
+	// snap-a was last refreshed yesterday
+	mockLastRefreshed(c, st, lastRefreshed, "snap-a")
+
+	// hold it for just a bit (10h) initially
+	hold := time.Hour * 10
+	c.Assert(snapstate.HoldRefresh(st, "snap-b", hold, "snap-a"), IsNil)
+	var gating map[string]map[string]*snapstate.HoldState
+	c.Assert(st.Get("snaps-hold", &gating), IsNil)
+	c.Check(gating, DeepEquals, map[string]map[string]*snapstate.HoldState{
+		"snap-a": {
+			"snap-b": snapstate.MockHoldState(now, "2021-05-10T20:00:00Z"),
+		},
+	})
+
+	// holding for a shorter time is fine too
+	hold = time.Hour * 5
+	c.Assert(snapstate.HoldRefresh(st, "snap-b", hold, "snap-a"), IsNil)
+	c.Assert(st.Get("snaps-hold", &gating), IsNil)
+	c.Check(gating, DeepEquals, map[string]map[string]*snapstate.HoldState{
+		"snap-a": {
+			"snap-b": snapstate.MockHoldState(now, "2021-05-10T15:00:00Z"),
+		},
+	})
+
+	oldNow := now
+
+	// a refresh on next day
+	now = "2021-05-11T08:00:00Z"
+
+	// default hold time requested
+	hold = 0
+	c.Assert(snapstate.HoldRefresh(st, "snap-b", hold, "snap-a"), IsNil)
+	c.Assert(st.Get("snaps-hold", &gating), IsNil)
+	c.Check(gating, DeepEquals, map[string]map[string]*snapstate.HoldState{
+		"snap-a": {
+			// maximum for holding other snaps, but taking into consideration
+			// firstHeld time = "2021-05-10T10:00:00".
+			"snap-b": snapstate.MockHoldState(oldNow, "2021-05-12T10:00:00Z"),
+		},
+	})
+}
+
+func (s *autorefreshGatingSuite) TestHoldRefreshHelperCloseToMaxPostponement(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	lastRefreshedStr := "2021-01-01T10:00:00Z"
+	lastRefreshed, err := time.Parse(time.RFC3339, lastRefreshedStr)
+	c.Assert(err, IsNil)
+	// we are 1 day before maxPostponent
+	now := lastRefreshed.Add(89 * time.Hour * 24)
+
+	restore := snapstate.MockTimeNow(func() time.Time { return now })
+	defer restore()
+
+	mockInstalledSnap(c, st, snapAyaml, false)
+	mockInstalledSnap(c, st, snapByaml, false)
+	mockLastRefreshed(c, st, lastRefreshedStr, "snap-a")
+
+	// request default hold time
+	var hold time.Duration
+	c.Assert(snapstate.HoldRefresh(st, "snap-b", hold, "snap-a"), IsNil)
+
+	var gating map[string]map[string]*snapstate.HoldState
+	c.Assert(st.Get("snaps-hold", &gating), IsNil)
+	c.Assert(gating, HasLen, 1)
+	c.Check(gating["snap-a"]["snap-b"].HoldUntil.String(), DeepEquals, lastRefreshed.Add(90*time.Hour*24).String())
+}
+
+func (s *autorefreshGatingSuite) TestHoldRefreshExplicitHoldTime(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	now := "2021-05-10T10:00:00Z"
+	restore := snapstate.MockTimeNow(func() time.Time {
+		t, err := time.Parse(time.RFC3339, now)
+		c.Assert(err, IsNil)
+		return t
+	})
+	defer restore()
+
+	mockInstalledSnap(c, st, snapAyaml, false)
+	mockInstalledSnap(c, st, snapByaml, false)
+
+	hold := time.Hour * 24 * 3
+	// holding self for 3 days
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", hold, "snap-a"), IsNil)
+
+	// snap-b holds snap-a for 1 day
```

Comment:
this comment is backwards right? in this call `snap-a` is holding `snap-b`? that's also what the map seems to show

Reply:
Thanks yes it makes sense now I think